In [1]:
import pandas as pd

In [3]:
df_train = pd.read_csv("df_train_merged.csv")
df_train.head()

Unnamed: 0  SK_ID_CURR  TARGET  NAME_CONTRACT_TYPE  CODE_GENDER  \
0           0      100002       1                   0            1   
1           1      100003       0                   0            0   
2           2      100004       0                   1            1   
3           3      100006       0                   0            0   
4           4      100007       0                   0            1   

   FLAG_OWN_CAR  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  \
0             0                1             0          202500.0    406597.5   
1             0                0             0          270000.0   1293502.5   
2             1                1             0           67500.0    135000.0   
3             0                1             0          135000.0    312682.5   
4             0                1             0          121500.0    513000.0   

   ...  Prev_CCB_CNT_INSTALMENT_MATURE_CUM  Prev_CCB_NAME_CONTRACT_STATUS  \
0  ...                                 NaN                            NaN   
1  ...                            6.666667                       0.064516   
2  ...                           12.782051                       0.000000   
3  ...                                 NaN                            NaN   
4  ...                            3.600000                       0.000000   

   Prev_CCB_SK_DPD  Prev_CCB_SK_DPD_DEF  Prev_POS_MONTHS_BALANCE  \
0              NaN                  NaN                -8.250000   
1              0.0                  0.0               -19.000000   
2              0.0                  0.0               -43.500000   
3              NaN                  NaN               -33.416667   
4              0.0                  0.0               -26.428571   

   Prev_POS_CNT_INSTALMENT  Prev_POS_CNT_INSTALMENT_FUTURE  \
0                 8.083333                        6.500000   
1                17.630769                       14.497436   
2                 8.500000                        4.916667   
3                12.400253                        6.796086   
4                11.333333                        8.111111   

   Prev_POS_NAME_CONTRACT_STATUS  Prev_POS_SK_DPD  Prev_POS_SK_DPD_DEF  
0                       1.785714         0.000000             0.000000  
1                       0.656410         0.000000             0.000000  
2                       0.333333         0.000000             0.000000  
3                       0.456805         0.000000             0.000000  
4                       0.461190         0.126984             0.126984  

[5 rows x 343 columns]

delete "Unnamed : 0" column , "Unnamed : 0"  column is always there every time we import exported data, make sure "Unnamed : 0" column is always be dropped

In [4]:
df_train = df_train.drop(["Unnamed: 0"], axis = 1)
df_train.head(5)

SK_ID_CURR  TARGET  NAME_CONTRACT_TYPE  CODE_GENDER  FLAG_OWN_CAR  \
0      100002       1                   0            1             0   
1      100003       0                   0            0             0   
2      100004       0                   1            1             1   
3      100006       0                   0            0             0   
4      100007       0                   0            1             0   

   FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0                1             0          202500.0    406597.5      24700.5   
1                0             0          270000.0   1293502.5      35698.5   
2                1             0           67500.0    135000.0       6750.0   
3                1             0          135000.0    312682.5      29686.5   
4                1             0          121500.0    513000.0      21865.5   

   ...  Prev_CCB_CNT_INSTALMENT_MATURE_CUM  Prev_CCB_NAME_CONTRACT_STATUS  \
0  ...                                 NaN                            NaN   
1  ...                            6.666667                       0.064516   
2  ...                           12.782051                       0.000000   
3  ...                                 NaN                            NaN   
4  ...                            3.600000                       0.000000   

   Prev_CCB_SK_DPD  Prev_CCB_SK_DPD_DEF  Prev_POS_MONTHS_BALANCE  \
0              NaN                  NaN                -8.250000   
1              0.0                  0.0               -19.000000   
2              0.0                  0.0               -43.500000   
3              NaN                  NaN               -33.416667   
4              0.0                  0.0               -26.428571   

   Prev_POS_CNT_INSTALMENT  Prev_POS_CNT_INSTALMENT_FUTURE  \
0                 8.083333                        6.500000   
1                17.630769                       14.497436   
2                 8.500000                        4.916667   
3                12.400253                        6.796086   
4                11.333333                        8.111111   

   Prev_POS_NAME_CONTRACT_STATUS  Prev_POS_SK_DPD  Prev_POS_SK_DPD_DEF  
0                       1.785714         0.000000             0.000000  
1                       0.656410         0.000000             0.000000  
2                       0.333333         0.000000             0.000000  
3                       0.456805         0.000000             0.000000  
4                       0.461190         0.126984             0.126984  

[5 rows x 342 columns]

In [5]:
df_train.shape

(307511, 342)

In [6]:
df_test = pd.read_csv("df_test_merged.csv")
df_test.head(5)

Unnamed: 0  SK_ID_CURR  NAME_CONTRACT_TYPE  CODE_GENDER  FLAG_OWN_CAR  \
0           0      100001                   0            0             0   
1           1      100005                   0            1             0   
2           2      100013                   0            1             1   
3           3      100028                   0            0             0   
4           4      100038                   0            1             1   

   FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0                1             0          135000.0    568800.0      20560.5   
1                1             0           99000.0    222768.0      17370.0   
2                1             0          202500.0    663264.0      69777.0   
3                1             2          315000.0   1575000.0      49018.5   
4                0             1          180000.0    625500.0      32067.0   

   ...  Prev_CCB_CNT_INSTALMENT_MATURE_CUM  Prev_CCB_NAME_CONTRACT_STATUS  \
0  ...                                 NaN                            NaN   
1  ...                            6.666667                       0.064516   
2  ...                           12.782051                       0.000000   
3  ...                                 NaN                            NaN   
4  ...                            3.600000                       0.000000   

   Prev_CCB_SK_DPD  Prev_CCB_SK_DPD_DEF  Prev_POS_MONTHS_BALANCE  \
0              NaN                  NaN                -8.250000   
1              0.0                  0.0               -19.000000   
2              0.0                  0.0               -43.500000   
3              NaN                  NaN               -33.416667   
4              0.0                  0.0               -26.428571   

   Prev_POS_CNT_INSTALMENT  Prev_POS_CNT_INSTALMENT_FUTURE  \
0                 8.083333                        6.500000   
1                17.630769                       14.497436   
2                 8.500000                        4.916667   
3                12.400253                        6.796086   
4                11.333333                        8.111111   

   Prev_POS_NAME_CONTRACT_STATUS  Prev_POS_SK_DPD  Prev_POS_SK_DPD_DEF  
0                       1.785714         0.000000             0.000000  
1                       0.656410         0.000000             0.000000  
2                       0.333333         0.000000             0.000000  
3                       0.456805         0.000000             0.000000  
4                       0.461190         0.126984             0.126984  

[5 rows x 342 columns]

In [7]:
df_test = df_test.drop(["Unnamed: 0"], axis = 1)
df_test.head(5)

SK_ID_CURR  NAME_CONTRACT_TYPE  CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  \
0      100001                   0            0             0                1   
1      100005                   0            1             0                1   
2      100013                   0            1             1                1   
3      100028                   0            0             0                1   
4      100038                   0            1             1                0   

   CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  AMT_GOODS_PRICE  \
0             0          135000.0    568800.0      20560.5         450000.0   
1             0           99000.0    222768.0      17370.0         180000.0   
2             0          202500.0    663264.0      69777.0         630000.0   
3             2          315000.0   1575000.0      49018.5        1575000.0   
4             1          180000.0    625500.0      32067.0         625500.0   

   ...  Prev_CCB_CNT_INSTALMENT_MATURE_CUM  Prev_CCB_NAME_CONTRACT_STATUS  \
0  ...                                 NaN                            NaN   
1  ...                            6.666667                       0.064516   
2  ...                           12.782051                       0.000000   
3  ...                                 NaN                            NaN   
4  ...                            3.600000                       0.000000   

   Prev_CCB_SK_DPD  Prev_CCB_SK_DPD_DEF  Prev_POS_MONTHS_BALANCE  \
0              NaN                  NaN                -8.250000   
1              0.0                  0.0               -19.000000   
2              0.0                  0.0               -43.500000   
3              NaN                  NaN               -33.416667   
4              0.0                  0.0               -26.428571   

   Prev_POS_CNT_INSTALMENT  Prev_POS_CNT_INSTALMENT_FUTURE  \
0                 8.083333                        6.500000   
1                17.630769                       14.497436   
2                 8.500000                        4.916667   
3                12.400253                        6.796086   
4                11.333333                        8.111111   

   Prev_POS_NAME_CONTRACT_STATUS  Prev_POS_SK_DPD  Prev_POS_SK_DPD_DEF  
0                       1.785714         0.000000             0.000000  
1                       0.656410         0.000000             0.000000  
2                       0.333333         0.000000             0.000000  
3                       0.456805         0.000000             0.000000  
4                       0.461190         0.126984             0.126984  

[5 rows x 341 columns]

In [8]:
df_test.shape

(48744, 341)

# Merge data train and test

Merge application test and application train. This function makes it easier to clean the data, can save time cleaning data on application tests and application trains at the same time

In [9]:
frames = [df_train,df_test]
df = pd.concat(frames)

In [10]:
df.head(5)

SK_ID_CURR  TARGET  NAME_CONTRACT_TYPE  CODE_GENDER  FLAG_OWN_CAR  \
0      100002     1.0                   0            1             0   
1      100003     0.0                   0            0             0   
2      100004     0.0                   1            1             1   
3      100006     0.0                   0            0             0   
4      100007     0.0                   0            1             0   

   FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0                1             0          202500.0    406597.5      24700.5   
1                0             0          270000.0   1293502.5      35698.5   
2                1             0           67500.0    135000.0       6750.0   
3                1             0          135000.0    312682.5      29686.5   
4                1             0          121500.0    513000.0      21865.5   

   ...  Prev_CCB_CNT_INSTALMENT_MATURE_CUM  Prev_CCB_NAME_CONTRACT_STATUS  \
0  ...                                 NaN                            NaN   
1  ...                            6.666667                       0.064516   
2  ...                           12.782051                       0.000000   
3  ...                                 NaN                            NaN   
4  ...                            3.600000                       0.000000   

   Prev_CCB_SK_DPD  Prev_CCB_SK_DPD_DEF  Prev_POS_MONTHS_BALANCE  \
0              NaN                  NaN                -8.250000   
1              0.0                  0.0               -19.000000   
2              0.0                  0.0               -43.500000   
3              NaN                  NaN               -33.416667   
4              0.0                  0.0               -26.428571   

   Prev_POS_CNT_INSTALMENT  Prev_POS_CNT_INSTALMENT_FUTURE  \
0                 8.083333                        6.500000   
1                17.630769                       14.497436   
2                 8.500000                        4.916667   
3                12.400253                        6.796086   
4                11.333333                        8.111111   

   Prev_POS_NAME_CONTRACT_STATUS  Prev_POS_SK_DPD  Prev_POS_SK_DPD_DEF  
0                       1.785714         0.000000             0.000000  
1                       0.656410         0.000000             0.000000  
2                       0.333333         0.000000             0.000000  
3                       0.456805         0.000000             0.000000  
4                       0.461190         0.126984             0.126984  

[5 rows x 342 columns]

In [11]:
df.shape

(356255, 342)

In [12]:
# Check the target value on df_train

df['TARGET'][:307511]

0         1.0
1         0.0
2         0.0
3         0.0
4         0.0
         ... 
307506    0.0
307507    0.0
307508    0.0
307509    1.0
307510    0.0
Name: TARGET, Length: 307511, dtype: float64

In [13]:
# Check the target value on df_test

df['TARGET'][307511:]

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
         ..
48739   NaN
48740   NaN
48741   NaN
48742   NaN
48743   NaN
Name: TARGET, Length: 48744, dtype: float64

In [14]:
# Drop the target column

df2 = df.drop(columns=['TARGET'])
df2.shape

(356255, 341)

Due to the application test on the target column all data contains a Nan value, therefore we drop the Target column, this aims to prevent the target column from entering the data cleaning process.

Redefine the column

# Handling Missing Value

check the missing values ​​for all features and make the proportions of the missing values ​​for each column and sort them from the most to the fewest missing value proportions

In [15]:
# Check the percentage of value Nan on All features

null = df2.isna().sum()/df2.shape[0]*100
null.sort_values(ascending=False)

Prev_FLAG_LAST_APPL_PER_CONTRACT                            100.000000
Prev_RATE_INTEREST_PRIVILEGED                                98.193710
Prev_RATE_INTEREST_PRIMARY                                   98.193710
Prev_CCB_AMT_PAYMENT_CURRENT                                 74.091592
Prev_CCB_AMT_DRAWINGS_POS_CURRENT                            74.078118
                                                               ...    
Prev_NAME_CASH_LOAN_PURPOSE_Repairs                           0.000000
Prev_NAME_CASH_LOAN_PURPOSE_Everyday expenses                 0.000000
Prev_NAME_CASH_LOAN_PURPOSE_Car repairs                       0.000000
Prev_NAME_CASH_LOAN_PURPOSE_Building a house or an annex      0.000000
Prev_WEEKDAY_APPR_PROCESS_START_THURSDAY                      0.000000
Length: 341, dtype: float64

In [16]:
# Showing top 10 features that has biggest Nan percentage

data_missing = pd.DataFrame(df2.isnull().sum() / len(df2)).sort_values(by=0, ascending=False)
data_missing[:10]

0
Prev_FLAG_LAST_APPL_PER_CONTRACT     1.000000
Prev_RATE_INTEREST_PRIVILEGED        0.981937
Prev_RATE_INTEREST_PRIMARY           0.981937
Prev_CCB_AMT_PAYMENT_CURRENT         0.740916
Prev_CCB_AMT_DRAWINGS_POS_CURRENT    0.740781
Prev_CCB_CNT_DRAWINGS_POS_CURRENT    0.740781
Prev_CCB_AMT_DRAWINGS_ATM_CURRENT    0.740781
Prev_CCB_AMT_DRAWINGS_OTHER_CURRENT  0.740781
Prev_CCB_CNT_DRAWINGS_ATM_CURRENT    0.740781
Prev_CCB_CNT_DRAWINGS_OTHER_CURRENT  0.740781

Base on the tabel above, feature on PREVIOUS_APPLICATION dataset has the most Nan Value

Drop features that has missing value more than 50%

In [17]:
# Droping features that has Value Nan > 50%

drop = data_missing[data_missing[0] > 0.5]
print(drop.count())
print(drop)

0    69
dtype: int64
                                          0
Prev_FLAG_LAST_APPL_PER_CONTRACT   1.000000
Prev_RATE_INTEREST_PRIVILEGED      0.981937
Prev_RATE_INTEREST_PRIMARY         0.981937
Prev_CCB_AMT_PAYMENT_CURRENT       0.740916
Prev_CCB_AMT_DRAWINGS_POS_CURRENT  0.740781
...                                     ...
APARTMENTS_AVG                     0.505110
APARTMENTS_MEDI                    0.505110
ENTRANCES_MEDI                     0.500785
ENTRANCES_MODE                     0.500785
ENTRANCES_AVG                      0.500785

[69 rows x 1 columns]


there are 69 features that have a missing value of more than 50%, so we drop those features

In [18]:
df2.drop(columns=drop.index, inplace=True)
df2.shape

(356255, 272)

The number of columns becomes 272 where previously it was 341.

In [19]:
df3 = df2.copy()
df3.head()

SK_ID_CURR  NAME_CONTRACT_TYPE  CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  \
0      100002                   0            1             0                1   
1      100003                   0            0             0                0   
2      100004                   1            1             1                1   
3      100006                   0            0             0                1   
4      100007                   0            1             0                1   

   CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  AMT_GOODS_PRICE  \
0             0          202500.0    406597.5      24700.5         351000.0   
1             0          270000.0   1293502.5      35698.5        1129500.0   
2             0           67500.0    135000.0       6750.0         135000.0   
3             0          135000.0    312682.5      29686.5         297000.0   
4             0          121500.0    513000.0      21865.5         513000.0   

   ...  Prev_IP_DAYS_INSTALMENT  Prev_IP_DAYS_ENTRY_PAYMENT  \
0  ...              -247.000000                 -254.800000   
1  ...              -533.122222                 -543.177778   
2  ...             -1294.198413                -1300.294709   
3  ...              -989.547348                 -998.087852   
4  ...              -700.413082                 -705.302259   

   Prev_IP_AMT_INSTALMENT  Prev_IP_AMT_PAYMENT  Prev_POS_MONTHS_BALANCE  \
0            21081.262500         21081.262500                -8.250000   
1            23136.942125         23136.942125               -19.000000   
2            11450.143036         11450.143036               -43.500000   
3            44632.814879         41626.031019               -33.416667   
4           102297.513891        101564.224500               -26.428571   

   Prev_POS_CNT_INSTALMENT  Prev_POS_CNT_INSTALMENT_FUTURE  \
0                 8.083333                        6.500000   
1                17.630769                       14.497436   
2                 8.500000                        4.916667   
3                12.400253                        6.796086   
4                11.333333                        8.111111   

   Prev_POS_NAME_CONTRACT_STATUS  Prev_POS_SK_DPD  Prev_POS_SK_DPD_DEF  
0                       1.785714         0.000000             0.000000  
1                       0.656410         0.000000             0.000000  
2                       0.333333         0.000000             0.000000  
3                       0.456805         0.000000             0.000000  
4                       0.461190         0.126984             0.126984  

[5 rows x 272 columns]

Redefine the dataframe

# Handling Missing Value

In [20]:
# fill the Nan Value with Median 

print("NAs for numerical features in train : " + str(df3.isnull().values.sum()))
df4 = df3.fillna(df3.median())
print("Remaining NAs for numerical features in train : " + str(df3.isnull().values.sum()))

NAs for numerical features in train : 2263233
Remaining NAs for numerical features in train : 2263233


In [21]:
# Recheck the Nan Value

sum(df4.isna().sum())/(df4.shape[0]*df4.shape[1])*100

0.0

After fill the missing value with median teh data has 0 Nan value

# Handling Variabel with Small Variances

Because small variance just give a little information

In [22]:
small_variance = df4.columns[(df4.std(axis = 0) < .01).values]
small_variance

Index(['FLAG_MOBIL', 'FLAG_DOCUMENT_2', 'FLAG_DOCUMENT_4', 'FLAG_DOCUMENT_10',
       'FLAG_DOCUMENT_12', 'Prev_NAME_CONTRACT_TYPE_XNA',
       'Prev_NAME_CASH_LOAN_PURPOSE_Buying a holiday home / land',
       'Prev_NAME_CASH_LOAN_PURPOSE_Buying a garage',
       'Prev_NAME_CASH_LOAN_PURPOSE_Business development',
       'Prev_NAME_CASH_LOAN_PURPOSE_Hobby',
       'Prev_NAME_CASH_LOAN_PURPOSE_Gasification / water supply',
       'Prev_NAME_CASH_LOAN_PURPOSE_Refusal to name the goal',
       'Prev_NAME_CASH_LOAN_PURPOSE_Money for a third person',
       'Prev_NAME_TYPE_SUITE_nan', 'Prev_NAME_GOODS_CATEGORY_Education',
       'Prev_NAME_GOODS_CATEGORY_Fitness',
       'Prev_NAME_GOODS_CATEGORY_Insurance',
       'Prev_NAME_GOODS_CATEGORY_Additional Service',
       'Prev_NAME_GOODS_CATEGORY_Weapon', 'Prev_NAME_GOODS_CATEGORY_Animals',
       'Prev_NAME_GOODS_CATEGORY_House Construction',
       'Prev_NAME_PORTFOLIO_Cars', 'Prev_CHANNEL_TYPE_Car dealer',
       'Prev_PRODUCT_COMBINATION_

In [23]:
df4.drop(columns = small_variance, inplace=True)

In [24]:
df4.shape

(356255, 248)

After handling Nan value and drop features that has small variace the columns on dataset become 248

# Recall the Target Feaatures

After clening the data, recall the column "TARGET" to dataframe

In [25]:
# Recall data target

df4['TARGET'] = df['TARGET']

In [26]:
df4.shape

(356255, 249)

In [27]:
# Check the Target Value on df_train

df4['TARGET'][:307511]

0         1.0
1         0.0
2         0.0
3         0.0
4         0.0
         ... 
307506    0.0
307507    0.0
307508    0.0
307509    1.0
307510    0.0
Name: TARGET, Length: 307511, dtype: float64

In [28]:
# Check the Target Value on df_test

df4['TARGET'][307511:]

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
         ..
48739   NaN
48740   NaN
48741   NaN
48742   NaN
48743   NaN
Name: TARGET, Length: 48744, dtype: float64

Target column in Application train and Test is back as before

# Check The Duplicate Data

In [29]:
print('The number of duplication is:', df4.duplicated().sum())

The number of duplication is: 0


There's no duplicate data

# Resplit the data to data Train and Test back

In [50]:
# Menimport data final

df4.to_csv("data_merged.csv")

In [51]:
# Split back data train and data test

df_train = df4[:307511]

In [52]:
df_train.shape

(307511, 275)

In [53]:
df_test = df4[307511:]

In [54]:
df_test.shape

(48744, 275)

# Export Data Train

In [55]:
# Eksport data final train

df_train_fix = df_train.copy()

In [56]:
df_train_fix.to_csv('df_train_fix.csv')

# Export Data Test

In [57]:
df_test_fix = df_test.copy()

In [58]:
df_test_fix.to_csv("df_test_fix.csv")